### PR TITLE
Backend Service

### PR DESCRIPTION
**Pull request by dgray001**
This PR adds the backend flask service.
This included adding a flask blueprint and using paramiko to connect to the ComputeEngine VM instance via its internal IP. Also use schlex.quote to prevent malicious scripts being sent through the input boxes.
This version can be accessed [here](https://20220216t091153-dot-personal-fa-starter-app.wl.r.appspot.com/client/).

Working UI:
image

Flask unit tests:
image

Note: no changes to frontend although git picks up a couple trivial changes as artifacts of rebasing.

### CODE REVIEW

File: submit.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+blueprint = flask.Blueprint('submit', __name__, url_prefix="/submit")
+
+blankData = {"inputString": "input", "inputFormat": "inputFormat", "outputFormat": "outputFormat", "additionalOptions": "flags", "output": "output", "log": ""}
+
+projectId = "personal-fa-starter-app"
+zone = "us-west4-b"
+
+@blueprint.route("/test/", methods=['PATCH'])
+def test():
```
Comment (katienguyen1) [inline submit.py]: Why is this called test?

File: submit.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    req = request.get_json()
+    print(req)
```
Comment (katienguyen1) [inline submit.py]: remove print statements

File: submit.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import os
+
+import flask
+from flask import request
+
+#import google.cloud_v1 as compute_v1
+
+blueprint = flask.Blueprint('submit', __name__, url_prefix="/submit")
+
+blankData = {"inputString": "input", "inputFormat": "inputFormat", "outputFormat": "outputFormat", "additionalOptions": "flags", "output": "output", "log": ""}
```
Comment (katienguyen1) [inline submit.py]: This should be deleted.

File: src/app/submit.service.ts [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
-import { mergeMap, catchError } from 'rxjs/operators';
+import { mergeMap, catchError, tap, delay } from 'rxjs/operators';
 import { OpenBabelData } from './OpenBabelData';
 
 @Injectable({
   providedIn: 'root'
 })
 export class SubmitService {
-  private readonly submitUrl: string = "/submit/";
+  private readonly submitUrl: string = "/submit/test";
```
Comment (katienguyen1) [inline src/app/submit.service.ts]: test is not appropriate route name.  perhaps you need to rename the route to be 'api/submit'

**Review by katienguyen1 — CHANGES_REQUESTED** [review]

File: src/app/submit.service.ts [hunk 5]
```diff
@@ -1,20 +1,34 @@
 import { Injectable } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
-import { Observable, of } from 'rxjs';
```
Comment (katienguyen1) [inline src/app/submit.service.ts]: All the changes you are making in this PR that's related to the service, you can move those changes to PR#9 ?

**Review by katienguyen1 — CHANGES_REQUESTED** [review]

File: submit.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+from shlex import quote
+
+blueprint = flask.Blueprint('submit', __name__, url_prefix="/submit")
+
+ip = "10.168.0.42"
+keyPath = "./obabel-key"
+
+@blueprint.route("/", methods=['PATCH'])
+def submit():
```
Comment (katienguyen1) [inline submit.py:16]: Can you look into making use of Mocks and add a test for this?  You could potentially create a mock object of the paramiko ssh client, then mock the return value of 'exec_command', (passing in whatever fake obabeCommand) and asserting the expected response from making an http post request with the submit route.

https://realpython.com/python-mock-library/#what-is-mocking
https://docs.python.org/3/library/unittest.mock.html

There are example usage here:
https://realpython.com/testing-third-party-apis-with-mocks/

**Review by katienguyen1 — CHANGES_REQUESTED** [review]

File: tests/test_app.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    def tearDown(self, client):
+        self.assertTrue(client is not None)
+        self.assertTrue(isinstance(client, FlaskClient))
+
+    def test_setup(self, client):
+        pass
+
+    def test_index_page(self, client):
+        rv: Response = client.get('/')
```
Comment (katienguyen1) [inline tests/test_app.py]: instead of abbreviating (rv), can you spell out for better readability. Change here and everywhere else.  Otherwise looks good.

**Review by katienguyen1 — CHANGES_REQUESTED** [review]

File: tests/test_app.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    def tearDown(self, client):
+        self.assertTrue(client is not None)
+        self.assertTrue(isinstance(client, FlaskClient))
+
+    def test_setup(self, client):
+        pass
+
+    def test_index_page(self, client):
+        response: Response = client.get('/')
+        self.assertEqual(rv.status_code, 200)
```
Comment (katienguyen1) [inline tests/test_app.py]: Did you rerun the tests and this still passes without having changed rv?

**Review by katienguyen1 — CHANGES_REQUESTED** [review]

File: tests/test_app.py (hunk 8 above)
Comment (dgray001) [reply tests/test_app.py]: Yes, changing the variable didn't make the tests fail. The unittest output looks identical so I didn't update the picture.

Comment (dgray001) [reply tests/test_app.py]: Oh wait yes I did make that typo for some reason the fix didn't get committed to what was pushed.

Comment (dgray001) [reply tests/test_app.py]: Yeh I messed up ci --amend -m somehow.

**Review by katienguyen1 — APPROVED** [review]

**Pull request merged**